third_party/nvfuser/python/__init__.py

```diff
@@ -2,26 +2,88 @@
 from . import _C
 from ._C import *
 
-class FusionDefinition(_C._FusionDefinition):
+import numbers
+from functools import partial
+
+
+class PythonOperators:
+    def _not_number_scalar(self, num):
+        return not isinstance(num, numbers.Number) or isinstance(num, bool)
+
+    def _convert_python_number(self, op, *args, **kwargs):
+        # TODO: handle python containers as well.
+        args = [
+            x if self._not_number_scalar(x) else self._fd.define_constant(x)
+            for x in args
+        ]
+        for k, v in kwargs.items():
+            if not self._not_number_scalar(v):
+                kwargs[k] = self._fd.define_constant(v)
+        return op(*args, **kwargs)
+
+    def bindOperation(self, op_name: str, op):
+        # TODO remove this after we can handle python containers.
+        if op_name not in ["var", "var_mean"]:
+            setattr(self, op_name, partial(self._convert_python_number, op))
+        else:
+            setattr(self, op_name, op)
+
+    def __init__(self, fd):
+        self._fd = fd
+
+
+class FusionDefinition:
+    Operators = _C._FusionDefinition.Operators
+
+    def __init__(self, id=None, *, max_length=None):
+        self._fd = _C._FusionDefinition(id)
+        self.ops = PythonOperators(self._fd)
+
+        # mapping Operators for query
+        self.Operator = self._fd.Operators
+
+        # map _C._FusionDefinition methods
+        map_methods = {
+            "_setup_definition": "setup_definition",
+            "_finalize_definition": "finalize_definition",
+            "id": "id",
+            "define_tensor": "define_tensor",
+            "define_scalar": "define_scalar",
+            "define_constant": "define_constant",
+            "add_output": "add_output",
+            "print": "print",
+            "__repr__": "__repr__",
+        }
+        for k, v in map_methods.items():
+            setattr(self, v, getattr(self._fd, k))
+
+        # map operator under ops
+        for op_name in filter(
+            lambda a: not a.startswith("_"), dir(_C._FusionDefinition.Operators)
+        ):
+            self.ops.bindOperation(op_name, getattr(self._fd.ops, op_name))
+
     def __enter__(self):
-        return self._setup_definition()
+        # note: do NOT return setup_definition()!
```

Comment:
Good catch! Until now, one would think that `with FusionDefinition() as fd:` is actually setting `fd` as a `_C.FusionDefinition`, but it seems we actually get the python object instead of the pybind11 base class. E.g. on devel:
```python
import nvfuser
fd = nvfuser.FusionDefinition()
with fd as fdin:
    print(type(fd), type(fdin), id(fd), id(fdin))
# <class 'nvfuser.FusionDefinition'> <class 'nvfuser.FusionDefinition'> 140482957092768 140482957092768
```
Still, as you put it, it's better to directly return self. You could remove the `_ =` or not.

Reply:
> You could remove the _ = or not.

makes sense~ I left the `_ =` when I want to show explicitly that we should not return the function. But a comment should suffice.

Reply:
> Until now, one would think that with FusionDefinition() as fd: is actually setting fd as a _C.FusionDefinition, but it seems we actually get the python object instead of the pybind11 base class. E.g. on devel:

Magic! I wonder if the change in behavior is a result of me changing the inheritance of `nvfuser.FusionDefinition`. But the existing behavior is puzzling...